### PR TITLE
1. separate timer in timer queue processor into dedicated file, add UT

### PR DESCRIPTION
**Pull request by wxing1292**
2. send visibility time through channel instead setting minPendingTimer
3. increase history cache size from 256 to 512, note, this value was previously 1024

### CODE REVIEW

**Comment by samarabbas** [comment]
Let's hold onto this change until we cut a new release.

File: service/history/timerQueueProcessor.go [hunk 1]
```diff
@@ -171,7 +87,7 @@ func newTimerQueueProcessor(shard ShardContext, historyService *historyEngineImp
 		cache:            historyService.historyCache,
 		executionManager: executionManager,
 		shutdownCh:       make(chan struct{}),
-		newTimerCh:       make(chan struct{}, 1),
+		newTimerCh:       make(chan time.Time, shard.GetConfig().TimerProcessorTimerChanSize),
```
Comment (samarabbas) [inline service/history/timerQueueProcessor.go]: we always do non blocking write to the channel.  Why do you need a buffer?

File: service/history/timerQueueProcessor.go [hunk 2]
```diff
@@ -238,16 +151,12 @@ func (t *timerQueueProcessorImpl) NotifyNewTimer(timerTasks []persistence.Task)
...
-	if updatedMinTimer {
-		select {
-		case t.newTimerCh <- struct{}{}:
-			// Notified about new timer.
-
-		default:
-			// Channel "full" -> drop and move on, since we are using it as an event.
-		}
+	select {
+	case t.newTimerCh <- newEarlistTime:
```
Comment (samarabbas) [inline service/history/timerQueueProcessor.go]: What happens if the earliest time is dropped due to non blocking write?

File: service/history/timer.go [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+package history
+
+import (
+	"sync"
+	"time"
+)
+
+type (
+	// Timer interface
+	Timer interface {
```
[the remark this answers is not in the document]
Comment (wxing1292) [reply service/history/timer.go]: ok

File: service/history/timer.go [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+		// GetTimerChan return the channel which will be fired when time is up
+		GetTimerChan() <-chan bool
+		// WillFireAfter check will the timer get fired after a certain time
+		WillFireAfter(now time.Time) bool
```
[the remark this answers is not in the document]
Comment (wxing1292) [reply service/history/timer.go]: changed to fireAfter. 
is after seems like comparing 2 time, and this function actually do "will this timer fire after time"

File: service/history/timer.go [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+import (
+	"sync"
+	"time"
+)
+
+type (
+	// Timer interface
+	Timer interface {
+		// GetTimerChan return the channel which will be fired when time is up
+		GetTimerChan() <-chan bool
```
[the remark this answers is not in the document]
Comment (wxing1292) [reply service/history/timer.go]: sure

File: service/history/timerQueueProcessor.go [hunk 6]
```diff
@@ -308,36 +217,31 @@ continueProcessor:
 				t.logger.Debug("Timer queue processor pump shutting down.")
 				return nil
 
-			case <-gateC:
+			case <-timerChan:
 				// Timer Fired.
 
-			case <-t.newTimerCh:
+			case newEarlistTime = <-t.newTimerCh:
```
[the remark this answers is not in the document]
Comment (wxing1292) [reply service/history/timerQueueProcessor.go]: sure

File: service/history/timerQueueProcessor.go (hunk 2 above)
Comment (wxing1292) [reply service/history/timerQueueProcessor.go]: changed back to use lock + variable for tracking min pending time

File: service/history/timerQueueProcessor.go [hunk 7]
```diff
@@ -217,32 +221,33 @@ continueProcessor:
...
-			}
-		}
 
-		if !newEarlistTime.IsZero() {
-			t.metricsClient.IncCounter(metrics.TimerQueueProcessorScope, metrics.NewTimerNotifyCounter)
-			t.logger.Debugf("Woke up by the timer")
+			case <-t.newTimerCh:
+				t.newTimeLock.Lock()
+				newTime := t.newTime
+				t.newTime = time.Time{}
```
Comment (samarabbas) [inline service/history/timerQueueProcessor.go]: can we define a null time const and set it to null time instead of creating a new struct each time.

Comment (wxing1292) [reply service/history/timerQueueProcessor.go]: const can only be character, string, boolean, or numeric values.

i can change the variable type to *time.Time

**Pull request merged**